### PR TITLE
upgrade to babel 6

### PR DESCRIPTION
**Pull request by erwinmombay**
Closes https://github.com/ampproject/amphtml/issues/9695

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,22 +1,12 @@
...
-    "es6.properties.shorthand",
-    "es6.templateLiterals",
-    "es6.spread",
-    "utility.inlineEnvironmentVariables"
-  ]
+  "presets": [
+    ["env", {
+      "loose": true,
+      "targets": {
+        "browsers": ["last 2 versions", "safari >= 7"]
```
Comment (erwinmombay) [inline .babelrc]: @jridgewell please correct this if needed

cc @cramforce

Comment (erwinmombay) [inline .babelrc]: @jridgewell im guessing you want this dynamic based on the original issue, depending on if we're in test or dev mode etc.

Comment (jridgewell) [inline .babelrc]: Yes, CI would get this string (`safari >= 9`, though is what I think we support), and in dev testing we would use the latest Chrome

File: extensions/amp-access/0.1/access-expr-impl.js [hunk 2]
```diff
@@ -175,7 +175,13 @@ parseError: function parseError(str, hash) {
...
         this.trace(str);
     } else {
-        throw new Error(str);
+        function _parseError (msg, hash) {
+            this.message = msg;
+            this.hash = hash;
+        }
+        _parseError.prototype = Error;
+
+        throw new _parseError(str, hash);
```
Comment (jridgewell) [inline extensions/amp-access/0.1/access-expr-impl.js]: What's this change for?

Comment (erwinmombay) [inline extensions/amp-access/0.1/access-expr-impl.js]: babel6 was erroring out on some label statements in the generated jison file, i tried upgrading jison and regenerated the files but it was still failing and had to add the "transform-remove-strict-mode" plugin to get it to work.

Comment (erwinmombay) [inline extensions/amp-access/0.1/access-expr-impl.js]: specifically its this line https://github.com/ampproject/amphtml/pull/9775/files#diff-8d20dae7ae84978e5dba004cfa21ff72L304

Comment (jridgewell) [inline extensions/amp-access/0.1/access-expr-impl.js]: Odd, that line shouldn't throw. 😞

File: gulpfile.js [hunk 3]
```diff
@@ -748,8 +748,20 @@ function compileJs(srcDir, srcFilename, destDir, options) {
         });
   }
 
+  var browsers = [
+    process.env.TRAVIS ? "safari >= 9" : "Last 1 Chrome versions"
```
Comment (jridgewell) [inline gulpfile.js]: This needs the "last 2 versions" in Travis.

**Comment by cramforce** [comment]
Nice!

**Comment by lannka** [comment]
I ended up into this issue when last time trying to do babel6: https://github.com/babel/babel/issues/4480
custom-element cries `illegal constructor bla blaa`

I suppose this has been fixed by one of the plugin?

**Comment by choumx** [comment]
Ah, so the [Jison error](https://travis-ci.org/ampproject/amphtml/jobs/270653166#L966) is caused by upgrading to 0.4.17. 

```
TypeError: Function.prototype.toString requires that 'this' be a Function
``` 

What happens that parse errors in Jison-generated files are changed from:

```js
throw new Error(str);
```

to...

```js
function _parseError (msg, hash) {
    this.message = msg;
    this.hash = hash;
}
_parseError.prototype = Error;

throw new _parseError(str, hash);
```

You can't call `toString()` on the latter. In Node and older versions of Chrome, the error is `TypeError: Function.prototype.toString is not generic`.

We can file a issue with Jison. Do we need to use 0.4.17 though?

**Comment by choumx** [comment]
He published a new release. 😄 Try upgrading to 0.4.18.

**Comment by erwinmombay** [comment]
a green build 😭

**Comment by erwinmombay** [comment]
@jridgewell can you give this a look

**Comment by erwinmombay** [comment]
will add the size.txt diff in a few.

File: test/size.txt [hunk 4]
```diff
@@ -1,98 +1,98 @@
       max   |         min   |        gzip   |   file
       ---   |         ---   |         ---   |   ---
-   290 kB   |    31.63 kB   |    12.49 kB   |   alp.js / alp.max.js
```
Comment (erwinmombay) [inline test/size.txt]: @jridgewell seems to be some heft increases on the babel sizes, lmk if theres any option that i missed turning on/off that would have done this.

it doesnt really affect the files we serve built from CC.

File: extensions/amp-viewer-integration/0.1/test/integration/test-amp-viewer-integration.js [hunk 5]
```diff
@@ -156,15 +156,15 @@ describes.sandboxed('AmpViewerIntegration', {}, () => {
       });
 
       const port = new WindowPortEmulator(
-        this.win, viewerOrigin);
+        window, viewerOrigin);
```
Comment (erwinmombay) [inline extensions/amp-viewer-integration/0.1/test/integration/test-amp-viewer-integration.js]: @chenshay can you double check that this is ok

File: package.json [hunk 6]
```diff
@@ -25,6 +25,7 @@
     "prepend-v0": "gulp prepend-global --prod node_modules/AMP_CONFIG.json --local --target dist/v0.js && gulp prepend-global --prod node_modules/AMP_CONFIG.json --local --target dist.3p/current-min/f.js"
   },
   "dependencies": {
+    "babel-polyfill": "6.26.0",
```
Comment (jridgewell) [inline package.json]: Hopefully this isn't needed.

File: test/size.txt (hunk 4 above)
Comment (jridgewell) [reply test/size.txt]: This is because of `babel-polyfill` which is **MASSIVE**

File: package.json (hunk 6 above)
Comment (erwinmombay) [reply package.json]: unfortunately it is needed for `regenerator-runtime`

Comment (jridgewell) [reply package.json]: I think regenerator only needs some symbol polyfills, which should be much smaller than the entire `babel-polyfill`.

Comment (erwinmombay) [reply package.json]: @jridgewell my expectation was `babel-preset-env` would have figured that out and added the polyfill but all the issues i found told me to include babel-polyfill. is there a specific package i can include that will allow me to remove babel-polyfill?

Comment (jridgewell) [reply package.json]: We can either drop regenerator (loose generator support in Safari 9), or dig through the regenerator runtime code to see what we need.

Comment (erwinmombay) [reply package.json]: @jridgewell got it. do you mind if i create a separate issue/pr for that?

Comment (jridgewell) [reply package.json]: Sure.

**Comment by erwinmombay** [comment]
follow up issue here https://github.com/ampproject/amphtml/issues/11291

**Comment by rsimha-amp** [comment]
@erwinmombay @danielrozenberg all yarn dependencies were removed in #11279. Looks like this PR wasn't rebased properly, and has re-added yarn.lock. It'll need to be removed again.

**Pull request merged**